File: Algorithm/Recursion/98.ts
// Definition for a binary tree node.
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function isValidBST(root: TreeNode | null): boolean {
  function validate(
    node: TreeNode | null,
    leftBound: number,
    rightBound: number
  ): boolean {
    if (!node) return true;
    if (!(leftBound < node.val && node.val < rightBound)) return false;
    return (
      validate(node.left, leftBound, node.val) &&
      validate(node.right, node.val, rightBound)
    );
  }
  return validate(root, -Infinity, Infinity);
}

File: Algorithm/Recursion/98.py
# Validate Binary Search Tree
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root):
        def validate(node, left_bound, right_bound):
            if not node:
                return True

            if not (left_bound < node.val and node.val < right_bound):
                return False

            return validate(node.left, left_bound, node.val) and validate(
                node.right, node.val, right_bound
            )

        return validate(root, float("-inf"), float("inf"))
